File: 2359 find closest node to given two nodes.cpp
class Solution {
public:
    int closestMeetingNode(vector<int>& edges, int node1, int node2)
    {
        vector<int> d1(edges.size(), 999999);
        vector<int> d2(edges.size(), 999999);
        int n = node1;
        //d1[node1] = 0;
        int cntr = 0;
        while(d1[n] == 999999)
        {
            d1[n] = cntr++;
            if(edges[n] == -1)
                break;
            n = edges[n];
        }
        //d2[node2] = 0;
        n = node2;
        cntr = 0;
        while(d2[n] == 999999)
        {
            d2[n] = cntr++;
            if(edges[n] == -1)
                break;
            n = edges[n];
        }
        int bestNode = -1;
        int bestDist = 999999;
        for(int i=0; i<edges.size(); ++i)
        {
            int m = max(d1[i], d2[i]);
            if(m < bestDist)
            {
                bestDist = m;
                bestNode = i;
            }
        }
        return bestNode;
    }
};

File: 2359 Find Closest Node to Given Two Nodes.py
class Solution:
    def closestMeetingNode(self, edges: List[int], node1: int, node2: int) -> int:
        if node1 == node2: return node1
        n = len(edges)
        distA = [-1] * n
        distA[node1] = 0
        distB = [-1] * n
        distB[node2] = 0
        i = node1
        j = node2
        k = 0
        while True:
            ret = []
            a = b = False
            if edges[i] != -1 and distA[edges[i]] == -1:
                i = edges[i]
                distA[i] = k
                if distB[i] != -1:
                    ret.append(i)
                a = True
            if edges[j] != -1 and distB[edges[j]] == -1:
                j = edges[j]
                distB[j] = k
                if distA[j] != -1:
                    ret.append(j)
                b = True
            if len(ret):
                return min(ret)
            k += 1
            if not a and not b:
                break
        return -1
